src/promotion/nhit/hash.c

```diff
@@ -0,0 +1,363 @@
+	struct nhit_hash *new_ctx;
+	uint64_t i;
+
+	new_ctx = env_vzalloc(sizeof(*new_ctx));
+	if (!new_ctx) {
+		result = -OCF_ERR_NO_MEM;
+		goto exit;
+	}
+
+	new_ctx->rb_entries = hash_size;
+	new_ctx->hash_entries = OCF_DIV_ROUND_UP(
+			new_ctx->rb_entries / 4,
+			HASH_PRIME) * HASH_PRIME - 1;
+
+	new_ctx->hash_map = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_map));
+	if (!new_ctx->hash_map) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_ctx;
+	}
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		new_ctx->hash_map[i] = new_ctx->rb_entries;
+
+	new_ctx->hash_locks = env_vzalloc(
+			new_ctx->hash_entries * sizeof(*new_ctx->hash_locks));
+	if (!new_ctx->hash_locks) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_hash;
+	}
+
+	for (i = 0; i < new_ctx->hash_entries; i++) {
+		if (env_rwsem_init(&new_ctx->hash_locks[i])) {
+			result = -OCF_ERR_UNKNOWN;
+			goto dealloc_locks;
+		}
+	}
+
+	new_ctx->ring_buffer = env_vzalloc(
+			new_ctx->rb_entries * sizeof(*new_ctx->ring_buffer));
+	if (!new_ctx->ring_buffer) {
+		result = -OCF_ERR_NO_MEM;
+		goto dealloc_locks;
+	}
+	for (i = 0; i < new_ctx->rb_entries; i++) {
+		new_ctx->ring_buffer[i].core_id = OCF_CORE_ID_INVALID;
+		env_atomic_set(&new_ctx->ring_buffer[i].counter, 0);
+	}
+
+	env_spinlock_init(&new_ctx->rb_pointer_lock);
+	new_ctx->rb_pointer = 0;
+
+	*ctx = new_ctx;
+	return 0;
+
+dealloc_locks:
+	env_vfree(new_ctx->hash_locks);
+	for (i = 0; i < new_ctx->hash_entries; i++)
+		ENV_BUG_ON(env_rwsem_destroy(&new_ctx->hash_locks[i]));
+dealloc_hash:
+	env_vfree(new_ctx->hash_map);
+dealloc_ctx:
+	env_vfree(new_ctx);
+exit:
+	return result;
+}
+
+void hash_deinit(nhit_hash_t ctx)
+{
+	ocf_cache_line_t i;
+
+	env_spinlock_destroy(&ctx->rb_pointer_lock);
+	for (i = 0; i < ctx->hash_entries; i++)
+		ENV_BUG_ON(env_rwsem_destroy(&ctx->hash_locks[i]));
+
+	env_vfree(ctx->ring_buffer);
+	env_vfree(ctx->hash_locks);
+	env_vfree(ctx->hash_map);
+	env_vfree(ctx);
+}
+
+static ocf_cache_line_t hash_function(ocf_core_id_t core_id, uint64_t core_lba,
+		uint64_t limit)
+{
+	if (core_id == OCF_CORE_ID_INVALID)
+		return limit;
+
+	return (ocf_cache_line_t) ((core_lba * HASH_PRIME + core_id) % limit);
+}
+
+static ocf_cache_line_t core_line_lookup(nhit_hash_t ctx,
+		ocf_core_id_t core_id, uint64_t core_lba)
+{
+	ocf_cache_line_t hash = hash_function(core_id, core_lba,
+			ctx->hash_entries);
+	ocf_cache_line_t needle = ctx->rb_entries;
+	ocf_cache_line_t cur;
+
+	if (ctx->hash_map[hash] == ctx->rb_entries)
+		return needle;
+
+	for (cur = ctx->hash_map[hash]; cur != ctx->rb_entries;
+			cur = ctx->ring_buffer[cur].coll_next) {
+		struct nhit_list_elem *cur_elem = &ctx->ring_buffer[cur];
+
+		if (cur_elem->core_lba == core_lba &&
+				cur_elem->core_id == core_id) {
+			needle = cur;
+			break;
+		}
+	}
+
+	return needle;
+}
+
+static inline uint64_t get_rb_slot(nhit_hash_t ctx)
+{
+	env_spinlock_lock(&ctx->rb_pointer_lock);
+
+	ctx->rb_pointer = (ctx->rb_pointer + 1) % ctx->rb_entries;
```

Comment:
Please store this value in some temporary variable, don't return it directly.

Reply:
Fixed